src/gapstate.c

```diff
@@ -5,30 +5,88 @@
  **
  ** This file declares all variables that are considered state for the
  ** interpreter
+ ** The background is that in HPC-GAP this state has to be thread local,
+ ** that is, an instance of the state has to exist for every thread, and
+ ** global variables are unsuitable for this purpose (even if protected
+ ** from concurrent access by locks).
  **
  */
 
 #include <src/system.h>
 #include <src/gapstate.h>
 
+static GAPStateHandler GAPStateHandlers[GAPSTATE_MAX_HANDLERS];
+static Int GAPStateHandlerCount = 0;
+static Int GAPStateNext = 0; // Start of next free memory area (as offset into GAPState.StateSlots)
+
 GAPState MainGAPState;
 
 void InitGAPState(GAPState * state)
 {
-    InitScannerState(state);
-    InitStatState(state);
-    InitExprState(state);
-    InitCoderState(state);
-    InitOpersState(state);
-
-    // RunConstructors?
+    RunGAPModuleConstructors();
 }
 
 void DestroyGAPState(GAPState * state)
 {
-    DestroyScannerState(state);
-    DestroyStatState(state);
-    DestroyExprState(state);
-    DestroyCoderState(state);
-    DestroyOpersState(state);
+    RunGAPModuleDestructors();
+}
+
+/*
+ * Register global state for a module.
+ *
+ * size - how much memory (in bytes) the module would like to reserve in
+ *        GAPState.StateSlots
+ * constructor and destructor - function pointers. These  functions are called
+ *        whenever a module is instantiiated. Currently in GAP this only happen
+ *        once at startup (and exit). In HPC-GAP this happens for every thread
+ *        creation and destruction.
+ *
+ * handle - if non-null, then the function writes the index into
+ *        GAPState.StateSlots that the module was
+ *        assigned.
+ *
+ * TODO: Should we align memory?
+ */
+void RegisterModuleState(UInt size, GAPModuleConstructor constructor, GAPModuleDestructor destructor, Int *handle)
```

Comment:
Logically, `size` and `handle` kind of belong together, don't they? So I'd move them to be together at the back: `constructor, destructor, size, handle`